### ISSUE TITLE
Unclear documentation on dynamic_rnn vs rnn for efficient dynamic sequence length computation

### ISSUE THREAD

**Issue by lingz** [tensorflow/tensorflow#3801]
It looks like from the latest documentation that `rnn` performs early stopping for dynamic length sequences whereas `dynamic_rnn` does not? This would seem to be the reverse of the intuition. 

So it looks like in commit 855d3b56014780a90143b3e0c0865334b188c2df, the definition of `dynamic_rnn` was changed from:

> The parameter `sequence_length` is required and dynamic calculation is
> automatically performed.

to:

>  The parameter `sequence_length` is optional and is used to copy-through state
>  and zero-out outputs when past a batch element's sequence length. So it's more
>  for correctness than performance, unlike in rnn().

Whereas the `rnn` documentation states:

>   If the sequence_length vector is provided, dynamic calculation is performed.
>   This method of calculation does not compute the RNN steps past the maximum
>   sequence length of the minibatch (thus saving computational time),

For me this makes it very unclear. In this context, does `rnn` do "more dynamic" unrolling than `dynamic_rnn` because it actually uses sequence_length to stop early. So if I want to do efficient variable length sequences dynamically, I should use `rnn` instead of `dynamic_rnn`. 

In this case, what does `dynamic_rnn` actually do (what is "dynamic unrolling" explicitly defined), except for accept input in a different format?

**Comment by ebrevdo** [comment]
This is correct.  When running `dynamic_rnn`, it is assumed that the inputs' max_time dimension is the **maximum number of frames in your minibatch**, and can vary from step to step.  In this case, early stopping is not required: we run up to the maximum number of timesteps in your minibatch.  In contrast when feeding `rnn` you must always provide a fixed number of inputs, which is in general longer than the maximum number of frames in your minibatch.  In this case `sequence_length` is also used to perform early stopping on the calculation.

That said, if you can suggest (or even better, submit a PR) with a clarifying comment, that would be great.

**Comment by alquraishi** [comment]
It would be nice if `dynamic_rnn` handled things like the fully unrolled `rnn`, by simply looking at the maximum length in `sequence_length` for each minibatch and quitting early. Requiring that the user actually vary the dimensions of the passed in tensor seems unnecessary, if the information contained in `sequence_length` is already sufficient to make the determination that past a certain time_step, all sequences in the current minibatch are padded and so it's safe to quit early.

**Comment by lingz** [comment]
But it seems like the earlier version of dynamic _rnn was documented to stop early rather than trying the full time dimension. Was there a reason for the change?

Also @ebrevdo it seems that you can change the max time steps from step to step, but you can't actually have different max time dimension within a single batch for dynamic_rnn. But for the regular rnn function you can optimize per sample per step. Is this correct?

**Comment by lingz** [comment]
Also. coul you clarify what it means to statically unroll an rnn vs dynamically unroll? The terms are a little unclear and I wasn't able to find much supporting documentation.

**Comment by ebrevdo** [comment]
@alquraishi it's wasteful (in terms of both memory usage *and* calls to `cond` within `dynamic_rnn`) to create a padded tensor past the max sequence length, and dynamic_rnn is trying to be as efficient as possible given constraints of wanting correctness.  since the user is able to change the shape from step to step, it is up to the user to do this correctly.

@lingz static unroll in `rnn` means you have a fixed number of input tensors and you run through all of them.  dynamic unroll in `rnn` means you stop calculating the RNNCell past the maximum sequence_length value, and fill in the rest of the outputs with zeros.

you cannot optimize calculations per example and per step; because tensorflow relies on vectorized calculations.  the best you can do is perform calculations up to the max sequence_length, zeroing out the outputs of examples that have finished as you go along,  then past the max sequence_length you can exit the calculation early as i explained in the previous paragraph.

long story short: early stopping in `rnn` relies on `sequence_length`, but is wasteful in `dynamic_rnn`.  `dynamic_rnn` relies on the fact that your input shape can vary from step to step to perform **true** dynamic stopping.  there is no per-example per-time dynamic stopping anywhere due to vectorization of operations.

if you would like to update the documentation clarifying these points, please submit a PR and I will be happy to review.  we can then reference this issue in the PR.
